CSHARP-685: Add metrics to the driver

joao-r-reis opened the pull request:
**JIRA**: [CSHARP-685](https://datastax-oss.atlassian.net/browse/CSHARP-685)

- [x] Implementation 
- [x] Unit / Integration tests
- [x] API Docs
- [x] Review
- [x] Benchmark
- [x] Metrics documentation section/page in driver's manual

## Builder API

```csharp
// App Metrics configuration
var metrics = 
    new MetricsBuilder()
            .Report.ToGraphite("http://127.0.0.1:2003")
            .Build();
var scheduler = new AppMetricsTaskScheduler(
    TimeSpan.FromSeconds(5),
    async () => { await Task.WhenAll(metrics.ReportRunner.RunAllAsync()); });

scheduler.Start();

// DataStax driver configuration
var cluster = 
    Cluster.Builder()
           .WithMetrics(metrics.CreateDriverMetricsProvider())
           .Build();

// DataStax driver configuration with advanced metrics settings
var cluster = 
    Cluster.Builder()
           .WithMetrics(
               metrics.CreateDriverMetricsProvider(),
               new MetricsOptions()
                   .SetEnabledNodeMetrics(new[] { NodeMetric.Meters.BytesSent })
                   .SetEnabledSessionMetrics(
                     SessionMetric.AllSessionMetrics.Except(new[] { SessionMetric.Meters.BytesSent }))
                   .SetBucketPrefix("web.app"))
           .Build();
```

## SessionMetric

```csharp
public sealed class SessionMetric : IMetric
    {
        public string Name { get; }

        public static readonly IEnumerable<SessionMetric> DefaultSessionMetrics = new[]
        {
            Meters.BytesSent,
            Meters.BytesReceived,

            Counters.CqlClientTimeouts,

            Gauges.ConnectedNodes
        };

        public static readonly IEnumerable<SessionMetric> AllSessionMetrics = 
             SessionMetric.DefaultSessionMetrics.Union(new[] { SessionMetric.Timers.CqlRequests });

        public static class Timers
        {
            public static readonly SessionMetric CqlRequests = new SessionMetric("cql-requests");
        }

        public static class Counters
        {
            public static readonly SessionMetric CqlClientTimeouts = new SessionMetric("cql-client-timeouts");
        }

        public static class Meters
        {
            public static readonly SessionMetric BytesSent = new SessionMetric("bytes-sent");

            public static readonly SessionMetric BytesReceived = new SessionMetric("bytes-received");
        }

        public static class Gauges
        {
            public static readonly SessionMetric ConnectedNodes = new SessionMetric("connected-nodes");
        }
    }
```

## NodeMetric

```csharp
public sealed class NodeMetric :  IMetric
    {
        public string Name { get; }

        public static readonly IEnumerable<NodeMetric> DefaultNodeMetrics = new[]
        {
            Meters.BytesSent,
            Meters.BytesReceived,

            Gauges.OpenConnections,
            Gauges.InFlight,
            
            Counters.SpeculativeExecutions,
            Counters.AuthenticationErrors,
            Counters.ConnectionInitErrors,

            Counters.AbortedRequests,
            Counters.OtherErrors,
            Counters.ReadTimeouts,
            Counters.UnavailableErrors,
            Counters.UnsentRequests,
            Counters.WriteTimeouts,
            Counters.ClientTimeouts,
            Counters.Errors,

            Counters.RetriesOnUnavailable,
            Counters.RetriesOnOtherError,
            Counters.RetriesOnReadTimeout,
            Counters.RetriesOnWriteTimeout,
            Counters.Retries,

            Counters.IgnoresOnReadTimeout,
            Counters.IgnoresOnUnavailable,
            Counters.IgnoresOnOtherError,
            Counters.IgnoresOnWriteTimeout,
            Counters.Ignores
        };

        public static readonly IEnumerable<NodeMetric> AllNodeMetrics = 
            NodeMetric.DefaultNodeMetrics.Union(new [] { NodeMetric.Timers.CqlMessages });

        public static class Counters
        {
            public static readonly NodeMetric SpeculativeExecutions = new NodeMetric("speculative-executions");

            public static readonly NodeMetric ConnectionInitErrors = new NodeMetric("errors.connection.init");

            public static readonly NodeMetric AuthenticationErrors = new NodeMetric("errors.connection.auth");

            public static readonly NodeMetric RetriesOnReadTimeout = new NodeMetric("retries.read-timeout");

            public static readonly NodeMetric RetriesOnWriteTimeout = new NodeMetric("retries.write-timeout");

            public static readonly NodeMetric RetriesOnUnavailable = new NodeMetric("retries.unavailable");

            public static readonly NodeMetric RetriesOnOtherError = new NodeMetric("retries.other");

            public static readonly NodeMetric Retries = new NodeMetric("retries.total");
            
            public static readonly NodeMetric IgnoresOnReadTimeout = new NodeMetric("ignores.read-timeout");

            public static readonly NodeMetric IgnoresOnWriteTimeout = new NodeMetric("ignores.write-timeout");

            public static readonly NodeMetric IgnoresOnUnavailable = new NodeMetric("ignores.unavailable");

            public static readonly NodeMetric IgnoresOnOtherError = new NodeMetric("ignores.other");

            public static readonly NodeMetric Ignores = new NodeMetric("ignores.total");
            
            public static readonly NodeMetric ReadTimeouts = new NodeMetric("errors.request.read-timeouts");

            public static readonly NodeMetric WriteTimeouts = new NodeMetric("errors.request.write-timeouts");

            public static readonly NodeMetric UnavailableErrors = new NodeMetric("errors.request.unavailables");

            public static readonly NodeMetric OtherErrors = new NodeMetric("errors.request.others");

            public static readonly NodeMetric AbortedRequests = new NodeMetric("errors.request.aborted");

            public static readonly NodeMetric UnsentRequests = new NodeMetric("errors.request.unsent");

            public static readonly NodeMetric ClientTimeouts = new NodeMetric("errors.request.client-timeouts");
        }

        public static class Gauges
        {
            public static readonly NodeMetric OpenConnections = new NodeMetric("pool.open-connections");

            public static readonly NodeMetric AvailableStreams = new NodeMetric("pool.available-streams");

            public static readonly NodeMetric InFlight = new NodeMetric("pool.in-flight");
        }

        public static class Timers
        {
            public static readonly NodeMetric CqlMessages = new NodeMetric("cql-messages");
        }

        public static class Meters
        {
            public static readonly NodeMetric BytesSent = new NodeMetric("bytes-sent");

            public static readonly NodeMetric BytesReceived = new NodeMetric("bytes-received");
        }
    }
```

## Custom Metric Providers API

```csharp
public interface IDriverMetricsProvider
{
    // Implementation of the several metric types;
    IDriverTimer Timer(string bucket, IMetric metric);

    IDriverMeter Meter(string bucket, IMetric metric);

    IDriverCounter Counter(string bucket, IMetric metric);

    IDriverGauge Gauge(string bucket, IMetric metric, Func<double?> valueProvider);

    void ShutdownMetricsBucket(string bucket);
}
```

### Interface for Counter metric (not showing others, this is just to show the thoughts behind the design)

```csharp
// Abstraction that the core driver uses
public interface IDriverCounter : IDriverMetric
{
    void Increment(long value);
}

// AppMetrics provider common interface for all metrics (lives on CassandraCSharpDriver.AppMetrics nuget package)
public interface IAppMetricsMetric : IDriverMetric
{
    string Context { get; }
    string Name { get; }
}

// AppMetrics based implementation of IDriverCounter on CassandraCSharpDriver.AppMetrics nuget package
public interface IAppMetricsCounter : IAppMetricsMetric, IDriverCounter
{
    long GetValue();
}
```

## Get Metrics API

```csharp
// Get metrics object
IDriverMetrics metrics = session.GetMetrics();

// Get session metrics and / or node metrics
IReadOnlyDictionary<Host, IMetricsRegistry<NodeMetric>> allNodeMetrics = metrics.NodeMetrics;
IMetricsRegistry<SessionMetric> sessionMetrics = metrics.SessionMetrics;

// Get specific host's node metrics
Host host = session.Cluster.AllHosts().First();
IMetricsRegistry nodeMetrics = allNodeMetrics[host];

// Get a specific metric of a specific host (requires AppMetrics Provider)
IAppMetricsCounter counterAppMetrics = 
    metrics.GetNodeMetric<IAppMetricsCounter>(host, NodeMetric.Counters.BytesSent);

// Get a specific metric of a specific host with AppMetrics extension method
IAppMetricsCounter counterAppMetrics = 
    metrics.GetNodeCounter(host, NodeMetric.Counters.BytesSent);

// Get the same metric without AppMetrics
IDriverCounter counter = 
    metrics.GetNodeMetric<IDriverCounter>(host, NodeMetric.Counters.BytesSent);

// Get value of that metric (AppMetricsCounter provides a GetValue() method)
long bytesSent = counterAppMetrics.GetValue();

public interface IDriverMetrics
{
    IMetricsRegistry<SessionMetric> SessionMetrics { get; }

    IReadOnlyDictionary<Host, IMetricsRegistry<NodeMetric>> NodeMetrics { get; }

    // TMetricType can be the driver interfaces like IDriverCounter or the type that the metrics provider returns (IAppMetricsCounter for our AppMetrics provider). This offers better support for custom implementations
    TMetricType GetNodeMetric<TMetricType>(Host host, NodeMetric nodeMetric) 
          where TMetricType : class, IDriverMetric;

    TMetricType GetSessionMetric<TMetricType>(SessionMetric sessionMetric) 
          where TMetricType : class, IDriverMetric;
}

// All metric types implement the following interface.
public interface IDriverMetric
{
}

// Metrics Registry has a collection for each metric type.
public interface IMetricsRegistry<TMetric> where TMetric : IMetric
{
    IReadOnlyDictionary<TMetric, IDriverCounter> Counters { get; }
    IReadOnlyDictionary<TMetric, IDriverGauge> Gauges { get; }
    IReadOnlyDictionary<TMetric, IDriverMeter> Meters { get; }
    IReadOnlyDictionary<TMetric, IDriverTimer> Timers { get; }
    IReadOnlyDictionary<TMetric, IDriverMetric> Metrics { get; }
}
```

## List of Metrics

### Implemented

#### Session Level

Name | Metric type
-- | --
cql-requests | Timer
cql-client-timeouts | Counter
bytes-sent | Meter
bytes-received | Meter
connected-nodes | Gauge

#### Node Level

Name | App.Metrics type
-- | --
speculative-executions | Counter
bytes-sent | Meter
bytes-received | Meter
cql-messages | Timer
errors.connection.init | Counter
errors.connection.auth | Counter
errors.request.read-timeouts | Counter
errors.request.write-timeouts | Counter
errors.request.unavailables | Counter
errors.request.others | Counter
errors.request.aborted | Counter
errors.request.unsent | Counter
errors.request.client-timeouts| Counter
pool.open-connections | Gauge
pool.in-flight | Gauge
retries.read-timeout | Counter
retries.write-timeout | Counter
retries.unavailable | Counter
retries.other | Counter
ignores.read-timeout | Counter
ignores.write-timeout | Counter
ignores.unavailable | Counter
ignores.other | Counter

joao-r-reis commented on src/Cassandra/Connections/Connection.cs:
```diff
@@ -555,14 +581,21 @@ internal bool ReadParse(byte[] buffer, int length)
-                var state = header.Opcode != EventResponse.OpCode
-                    ? RemoveFromPending(header.StreamId)
-                    : new OperationState(EventHandler);
+                // todo (sivukhin, 14.04.2019): Is it crucial to remove state from pending operations strictly before writing remaining body to the stream?
```
should remove from pending asap

joao-r-reis commented on src/Cassandra/Connections/Connection.cs:
```diff
@@ -732,11 +765,12 @@ public OperationState Send(IRequest request, Action<Exception, Response> callbac
+            var state = new OperationState(
+                callback,
+                request,
+                timeoutMillis > 0 ? timeoutMillis : Configuration.SocketOptions.ReadTimeoutMillis,
```
This is not needed (it was changed on `master` so this "decision" is made before this point)

joao-r-reis commented on src/Cassandra/Connections/Connection.cs:
```diff
@@ -822,9 +857,6 @@ private void RunWriteQueueAction()
                     break;
                 }
-                //We will not use the request any more, stop reference it.
-                state.Request = null;
```
add this back
edit: this wasn't removed, just moved inside `WriteFrame`

joao-r-reis commented on src/Cassandra/IInternalStatement.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace Cassandra
+{
+    internal interface IInternalStatement : IStatement
```
internal xml docs

joao-r-reis commented on src/Cassandra/ISession.cs:
```diff
@@ -62,6 +63,8 @@ public interface ISession: IDisposable
         /// </summary>
         UdtMappingDefinitions UserDefinedTypes { get; }
 
+        string SessionName { get; }
```
xml docs

joao-r-reis commented on src/Cassandra/ISession.cs:
```diff
@@ -215,6 +218,8 @@ public interface ISession: IDisposable
         /// <param name="customPayload">Custom outgoing payload to send with the prepare request</param>
         Task<PreparedStatement> PrepareAsync(string cqlQuery, IDictionary<string, byte[]> customPayload);
 
+        IDriverMetrics GetMetrics();
```
xml docs

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IDriverCounter.cs:
```diff
@@ -0,0 +1,27 @@
+// 
+//       Copyright (C) 2019 DataStax Inc.
```
remove dates in license header

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IDriverCounter.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace Cassandra.Metrics.Abstractions
+{
+    public interface IDriverCounter : IDriverMetric<long>
```
xml docs

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IDriverGauge.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace Cassandra.Metrics.Abstractions
+{
+    public interface IDriverGauge : IDriverMetric<double?>, IDisposable
```
xml docs

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IDriverHistogram.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace Cassandra.Metrics.Abstractions
+{
+    public interface IDriverHistogram : IDriverMetric<IHistogramValue>
```
xml docs

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IDriverMeter.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace Cassandra.Metrics.Abstractions
+{
+    public interface IDriverMeter : IDriverMetric<IMeterValue>
```
xml docs

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IDriverMetric`.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace Cassandra.Metrics.Abstractions
+{
+    public interface IDriverMetric<out T> : IDriverMetric
```
should be `internal`?

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IDriverMetric.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace Cassandra.Metrics.Abstractions
+{
+    public interface IDriverMetric
```
xml docs

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IDriverMetricsProvider.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace Cassandra.Metrics.Abstractions
+{
+    public interface IDriverMetricsProvider
```
xml docs

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IDriverTimeHandler.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace Cassandra.Metrics.Abstractions
+{
+    public interface IDriverTimeHandler
```
xml docs and maybe change name? `IDriverTimeMeasurement`?

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IDriverTimeHandler.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    public interface IDriverTimeHandler
+    {
+        void EndRecording();
```
`StopMeasuring`?

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IDriverTimer.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace Cassandra.Metrics.Abstractions
+{
+    public interface IDriverTimer : IDriverMetric<ITimerValue>
```
xml docs

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IDriverTimer.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    public interface IDriverTimer : IDriverMetric<ITimerValue>
+    {
+        IDriverTimeHandler StartRecording();
```
`StartMeasuring`?

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IHistogramValue.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace Cassandra.Metrics.Abstractions
+{
+    public interface IHistogramValue
```
xml docs

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IMeterValue.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace Cassandra.Metrics.Abstractions
+{
+    public interface IMeterValue
```
xml docs

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/ITimerValue.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace Cassandra.Metrics.Abstractions
+{
+    public interface ITimerValue
```
xml docs

joao-r-reis commented on src/Cassandra/Metrics/IDriverMetrics.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace Cassandra.Metrics
+{
+    public interface IDriverMetrics
```
xml docs

joao-r-reis commented on src/Cassandra/Metrics/IDriverMetrics.cs:
```diff
@@ -0,0 +1,26 @@
+    {
+        IMetricsRegistry SessionMetrics { get; }
+
+        IReadOnlyDictionary<Host, IMetricsRegistry> NodeMetrics { get; }
```
test if entry is removed from this dictionary when `Host` is removed

joao-r-reis commented on src/Cassandra/Metrics/IMetricsRegistry.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace Cassandra.Metrics
+{
+    public interface IMetricsRegistry
```
xml docs

joao-r-reis commented on src/Cassandra/Metrics/Internal/IMetricsManager.cs:
```diff
@@ -0,0 +1,33 @@
+
+namespace Cassandra.Metrics.Internal
+{
+    internal interface IMetricsManager : IDriverMetrics, IDisposable
```
internal xml docs

joao-r-reis commented on src/Cassandra/Metrics/Providers/AppMetrics/AppMetricsDriverCounter.cs:
```diff
@@ -0,0 +1,68 @@
+
+using Cassandra.Metrics.Abstractions;
+
+namespace Cassandra.Metrics.Providers.AppMetrics
```
move everything on this namespace to another package

joao-r-reis commented on src/Cassandra/Metrics/Providers/AppMetrics/AppMetricsDriverGauge.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace Cassandra.Metrics.Providers.AppMetrics
+{
+    // todo(sivukhin, 07.08.2019): Dispose gauges somehow?
```
this is implemented, add tests and remove TODO

joao-r-reis commented on src/Cassandra/Metrics/Registries/INodeMetrics.cs:
```diff
@@ -0,0 +1,52 @@
+
+namespace Cassandra.Metrics.Registries
+{
+    internal interface INodeMetrics : IMetricsRegistry, IDisposable
```
internal xml docs

joao-r-reis commented on src/Cassandra/Metrics/Registries/IRequestMetrics.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace Cassandra.Metrics.Registries
+{
+    internal interface IRequestMetrics
```
internal xml docs

joao-r-reis commented on src/Cassandra/Metrics/Registries/ISessionMetrics.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace Cassandra.Metrics.Registries
+{
+    internal interface ISessionMetrics : IMetricsRegistry, IDisposable
```
internal xml docs

joao-r-reis commented on src/Cassandra/Metrics/Registries/SessionMetricsRegistry.cs:
```diff
@@ -0,0 +1,81 @@
+                BytesReceived = _driverMetricsProvider.Counter("bytes-received", DriverMeasurementUnit.Bytes);
+                ConnectedNodes = _driverMetricsProvider.Gauge(
+                    "connected-nodes",
+                    () => session.GetPools().Count(), //TODO
```
add internal `Counter` similar to `CountableConcurrentStack`

edit: `session.CountAllConnections` seems to be enough after changing `ConcurrentDictionary` to `CopyOnWriteDictionary`

joao-r-reis commented on src/Cassandra/MetricsOptions.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace Cassandra
+{
+    public class MetricsOptions
```
xml docs

joao-r-reis commented on src/Cassandra/NodeMetrics.cs:
```diff
@@ -0,0 +1,156 @@
+
+namespace Cassandra
+{
+    public static class NodeMetrics
```
xml docs

joao-r-reis commented on src/Cassandra/Observers/Abstractions/IConnectionObserver.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace Cassandra.Observers.Abstractions
+{
+    internal interface IConnectionObserver
```
internal xml docs

joao-r-reis commented on src/Cassandra/Observers/Abstractions/IHostObserver.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace Cassandra.Observers.Abstractions
+{
+    //TODO DELETE
```
reminder

joao-r-reis commented on src/Cassandra/Observers/Abstractions/IObserverFactory.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace Cassandra.Observers.Abstractions
+{
+    internal interface IObserverFactory
```
internal xml docs

joao-r-reis commented on src/Cassandra/Observers/Abstractions/IOperationObserver.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace Cassandra.Observers.Abstractions
+{
+    internal interface IOperationObserver
```
internal xml docs

joao-r-reis commented on src/Cassandra/Observers/Abstractions/IRequestObserver.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace Cassandra.Observers.Abstractions
+{
+    internal interface IRequestObserver
```
internal xml docs

joao-r-reis commented on src/Cassandra/Observers/Abstractions/ISessionObserver.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace Cassandra.Observers.Abstractions
+{
+    //TODO DELETE
```
reminder

joao-r-reis commented on src/Cassandra/Observers/HostObserver.cs:
```diff
@@ -0,0 +1,87 @@
+
+namespace Cassandra.Observers
+{
+    //TODO DELETE
```
reminder

joao-r-reis commented on src/Cassandra/Observers/IObserverFactoryBuilder.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace Cassandra.Observers
+{
+    internal interface IObserverFactoryBuilder
```
internal xml docs

joao-r-reis commented on src/Cassandra/Observers/NullConnectionObserver.cs:
```diff
@@ -0,0 +1,47 @@
+
+namespace Cassandra.Observers
+{
+    //TODO DELETE??
```
reminder to check if this is used anywhere

joao-r-reis commented on src/Cassandra/Observers/NullOperationObserver.cs:
```diff
@@ -0,0 +1,33 @@
+
+namespace Cassandra.Observers
+{
+    //TODO DELETE??
```
reminder to check if this is used anywhere

joao-r-reis commented on src/Cassandra/Observers/NullRequestObserver.cs:
```diff
@@ -0,0 +1,45 @@
+
+namespace Cassandra.Observers
+{
+    internal class NullRequestObserver : IRequestObserver
```
reminder to check if this is used anywhere

joao-r-reis commented on src/Cassandra/Observers/OperationObserver.cs:
```diff
@@ -0,0 +1,69 @@
+
+        private static void LogError(Exception ex)
+        {
+            Logger.Warning("An error occured while recording metrics for a connection operation. Exception = {0}", ex.ToString());
```
Should this be a warning?

joao-r-reis replied:
cc @jorgebay

joao-r-reis commented on src/Cassandra/Observers/RequestObserver.cs:
```diff
@@ -0,0 +1,121 @@
+
+        private void OnRequestRetry(IRequestMetrics metricsRegistry, RetryReasonType reason)
+        {
+            // todo(sivukhin, 08.08.2019): Missed OnAborted metric?
```
reminder to check what the java driver uses to record this metric

joao-r-reis replied:
In the context of the java driver, abort is an error where it was not possible to obtain the server response at all: https://github.com/datastax/java-driver/blob/9f0d89799a8a1e4cd1022dd7c43333924c36a648/core/src/main/java/com/datastax/oss/driver/internal/core/channel/ResponseCallback.java#L36-L47

joao-r-reis commented on src/Cassandra/Observers/SessionObserver.cs:
```diff
@@ -0,0 +1,72 @@
+
+namespace Cassandra.Observers
+{
+    //TODO DELETE??
```
reminder

joao-r-reis commented on src/Cassandra/Policies/RetryReasonType.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace Cassandra
+{
+    public enum RetryReasonType
```
can be `internal`?

joao-r-reis commented on src/Cassandra/Requests/IRequestExecutionFactory.cs:
```diff
@@ -14,12 +14,14 @@
 
 namespace Cassandra.Requests
 {
     internal interface IRequestExecutionFactory
```
internal xml docs

joao-r-reis commented on src/Cassandra/Requests/IRequestResultHandler.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace Cassandra.Requests
+{
+    internal interface IRequestResultHandler
```
internal xml docs

joao-r-reis commented on src/Cassandra/SessionMetrics.cs:
```diff
@@ -0,0 +1,45 @@
+
+namespace Cassandra
+{
+    public static class SessionMetrics
```
xml docs

joao-r-reis commented on src/Cassandra/Metrics/Registries/NodeMetricsRegistry.cs:
```diff
@@ -0,0 +1,119 @@
+            _hostConnectionPool = pool;
+        }
+
+        //TODO
```
delete

joao-r-reis commented on src/Cassandra/Metrics/Registries/NodeMetricsRegistry.cs:
```diff
@@ -0,0 +1,119 @@
+            Gauges = new[] { OpenConnections, AvailableStreams, InFlight, MaxRequestsPerConnection };
+            Meters = Errors.Meters.Concat(Ignores.Meters).Concat(Retries.Meters);
+            Timers = new[] { CqlMessages };
+        }
```
initialize disabled metrics with `Null` implementations

joao-r-reis commented on src/Cassandra/Metrics/Registries/SessionMetricsRegistry.cs:
```diff
@@ -0,0 +1,81 @@
+                ConnectedNodes?.Dispose();
+                throw;
+            }
+        }
```
initialize disabled metrics with `Null` Implementations

joao-r-reis commented on src/Cassandra/Metrics/Registries/RequestMetricsRegistry.cs:
```diff
@@ -0,0 +1,49 @@
+            Unavailable = driverMetricsProvider.Meter("unavailables", DriverMeasurementUnit.Requests);
+            Other = driverMetricsProvider.Meter("other", DriverMeasurementUnit.Requests);
+            Total = driverMetricsProvider.Meter("total", DriverMeasurementUnit.Requests);
+        }
```
Initialize disabled metrics with `Null` implementations

joao-r-reis commented on src/Cassandra/Metrics/Registries/RequestMetricsRegistry.cs:
```diff
@@ -0,0 +1,49 @@
+            Total = driverMetricsProvider.Meter("total", DriverMeasurementUnit.Requests);
+        }
+
+        public IEnumerable<IDriverMeter> Meters => new[] { Aborted, ReadTimeout, WriteTimeout, Unavailable, Other };
```
Create array once after initialization

joao-r-reis commented on src/Cassandra/Metrics/Providers/Null/NullDriverCounter.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace Cassandra.Metrics.Providers.Null
+{
+    internal class NullDriverCounter : IDriverCounter
```
Hmm I named these `Null`  as a reference to the null object design pattern but if you think this might confuse other devs that look at the code I'll change it to `Empty` which should be clearer.

jorgebay replied:
bah, makes sense to continue using `Null`.

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/DriverMeasurementUnit.cs:
```diff
@@ -0,0 +1,27 @@
+// 
+//       Copyright (C) 2019 DataStax Inc.
```
Yeah I forgot about this, I'll run a simple console app that I have just for this operation tomorrow morning

joao-r-reis commented on src/Extensions/Cassandra.AppMetrics/Cassandra.AppMetrics.csproj:
```diff
@@ -0,0 +1,36 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\Cassandra\Cassandra.csproj" />
```
Here is the generated nuspec inside the nupkg returned by `dotnet pack Cassandra.AppMetrics.csproj`:

```
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2013/05/nuspec.xsd">
  <metadata>
    <id>CassandraCSharpDriver.AppMetrics</id>
    <version>1.0.0-alpha2</version>
    <title>DataStax C# Drivers App Metrics Extension</title>
    <authors>DataStax</authors>
    <owners>DataStax</owners>
    <requireLicenseAcceptance>false</requireLicenseAcceptance>
    <licenseUrl>https://github.com/datastax/csharp-driver/blob/master/LICENSE</licenseUrl>
    <projectUrl>https://github.com/datastax/csharp-driver</projectUrl>
    <description>This package builds on the DataStax Enterprise C# driver and DataStax C# Driver for Apache Cassandra, adding a metrics provider implementation using App Metrics.</description>
    <copyright>Copyright © by DataStax</copyright>
    <tags>cassandra apache datastax dse driver client metrics appmetricsdatabase nosql dotnet netcore db</tags>
    <repository url="https://github.com/datastax/csharp-driver" />
    <dependencies>
      <group targetFramework=".NETFramework4.6.1">
        <dependency id="CassandraCSharpDriver" version="3.11.0" exclude="Build,Analyzers" />
        <dependency id="App.Metrics.Abstractions" version="3.1.0" exclude="Build,Analyzers" />
      </group>
      <group targetFramework=".NETStandard2.0">
        <dependency id="CassandraCSharpDriver" version="3.11.0" exclude="Build,Analyzers" />
        <dependency id="App.Metrics.Abstractions" version="3.1.0" exclude="Build,Analyzers" />
      </group>
    </dependencies>
  </metadata>
</package>
```

It looks like it is correctly picking up the dependency as a package dependency. If I'm not mistaken this was one of the features of the new `dotnet` CLI when .NET Core came out, i.e., treating every project as a package.

jorgebay replied:
Would it be possible to use version ranges?

It would be nice if we could support multiple `CassandraCSharpDriver` versions from a single `CassandraCSharpDriver.AppMetrics` version, avoiding to publish one `CassandraCSharpDriver.AppMetrics` version per `CassandraCSharpDriver` version.

Is it possible to support version ranges of `CassandraCSharpDriver`?

joao-r-reis replied:
So by default `CassandraCSharpDriver.AppMetrics` will require `CassandraCSharpDriver` >= 3.11.0 assuming 3.11.0 is the version of `CassandraCSharpDriver`  at the time when `dotnet pack` is run for `CassandraCSharpDriver.AppMetrics`. With this we wouldn't have to build `CassandraCSharpDriver.AppMetrics` everytime but it does mean that whenever we build `CassandraCSharpDriver.AppMetrics` for a new release we are bumping the dependency version requirement. 

Are you talking about manually setting the version requirement like  `CassandraCSharpDriver` >= 3.0 and < 4.0? This is doable but with a manually created `.nuspec` only (the old way)

joao-r-reis commented on src/Cassandra/Collections/CopyOnWriteDictionary.cs:
```diff
@@ -83,6 +75,12 @@ public bool ContainsKey(TKey key)
         public TValue GetOrAdd(TKey key, TValue value)
         {
+            // optimistic scenario: return before lock
+            if (_map.TryGetValue(key, out var outPutValue))
```
Agree, done

joao-r-reis commented on src/Cassandra/Connections/HostConnectionPool.cs:
```diff
@@ -103,17 +110,20 @@ private static class PoolState
         public IConnection[] ConnectionsSnapshot => _connections.GetSnapshot();
 
-        public HostConnectionPool(Host host, Configuration config, Serializer serializer)
+        public Host Host { get; private set; }
```
changed to field

joao-r-reis commented on src/Cassandra/Metrics/Abstractions/IDriverMeter.cs:
```diff
@@ -0,0 +1,28 @@
+        /// <summary>
+        /// Same as calling <see cref="Mark(long)"/> with amount=1.
+        /// </summary>
+        void Mark();
```
good catch, I recently changed most meters to counters and this doesn't make sense anymore

joao-r-reis commented on src/Extensions/Cassandra.AppMetrics/Cassandra.AppMetrics.csproj:
```diff
@@ -0,0 +1,36 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="App.Metrics.Abstractions" Version="[3.0.0,4.0)" />
```
changed to `3.*` (TIL)

joao-r-reis commented on src/Cassandra/Cluster.cs:
```diff
@@ -242,13 +245,50 @@ public async Task<ISession> ConnectAsync(string keyspace)
+
+        private async Task<long> GetAndIncrementSessionCounterAsync()
+        {
+            await _sessionCounterLock.WaitAsync().ConfigureAwait(false);
```
This might be a bit nitpicky but I'm concerned that someone out there might be constantly creating sessions from the same cluster instance so I'd rather not impose a set max number of sessions. I'm +1 on removing locks though. I've changed this, lmk your thoughts.

jorgebay replied:
lgtm!

jorgebay commented on src/Cassandra/Cluster.cs:
```diff
@@ -242,13 +244,32 @@ public async Task<ISession> ConnectAsync(string keyspace)
+
+        private long GetAndIncrementSessionCounter()
+        {
+            var newCounter = Interlocked.Increment(ref _sessionCounter);
```
Is this because of overflow?
If so, we could add a code comment.

joao-r-reis replied:
Yeah just to avoid negative counters if it overflows, I'll add a comment

jorgebay commented on src/Cassandra/Connections/CountableConcurrentStack.cs:
```diff
@@ -0,0 +1,60 @@
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Cassandra.Connections
```
We should move it to `Collections` namespace.

jorgebay commented on src/Cassandra/Metrics/Registries/NodeMetrics.cs:
```diff
@@ -0,0 +1,108 @@
+                    _context, NodeMetric.Gauges.OpenConnections, () => _hostConnectionPool?.OpenConnections, DriverMeasurementUnit.None);
+
+                AvailableStreams = MetricsRegistry.Gauge(
+                    _context, NodeMetric.Gauges.AvailableStreams, () => _hostConnectionPool?.AvailableStreams, DriverMeasurementUnit.None);
```
Is this necessary?
I think it doesn't give more info than `InFlight` and `OpenConnections` and it requires significant overhead to maintain it (2 interlocked ops per execution).

joao-r-reis replied:
That's a good point, I thought this would be useful as the java driver also has it, maybe for scenarios of very high throughput where it might be nice to see how close the connection is to being saturated? I don't know how much impact these 2 interlocked ops have on each execution, I still haven't started coding and running some benchmarks.

joao-r-reis replied:
We can always choose not to add this metric and property now. We can always implement it later if there is demand for it, wdyt

jorgebay replied:
Yeah, I'm +1 on removing it now, we can add it later if needed.

About op cost, there's a good estimation here: http://ithare.com/infographics-operation-costs-in-cpu-clock-cycles/

joao-r-reis replied:
Thanks for link, interesting stuff 👍 
Deleted `CountableConcurrentStack` and `AvailableStreams` metric

jorgebay commented on src/Cassandra/Connections/Connection.cs:
```diff
@@ -117,6 +119,8 @@ internal class Connection : IConnection
         /// </summary>
         public virtual int InFlight => Volatile.Read(ref _inFlight);
 
+        public virtual int AvailableStreams => _freeOperations.Count;
```
I think we should kill this property (see gauge discussion)

jorgebay commented on src/Cassandra/IInternalStatement.cs:
```diff
@@ -0,0 +1,31 @@
+    /// <summary>
+    /// Internal interface that exposes useful methods and properties for the driver internal components.
+    /// </summary>
+    internal interface IInternalStatement : IStatement
```
I think we should move this to `Data.Linq` namespace.

jorgebay commented on src/Cassandra/IMetric.cs:
```diff
@@ -0,0 +1,28 @@
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+namespace Cassandra
```
I've noticed that some part of the API is on the root namespace, is this intentional?

I think we can expose them in `Metrics` namespace, like `IDriverMetrics` and `IMetricsRegistry`.

joao-r-reis replied:
I tried to have everything related to the `Builder` on the root namespace so that when users try to call `builder.WithMetrics(root.CreateDriverMetricsProvider(), new MetricOptions())`, they wouldn't run into the namespace issue. This means that `NodeMetric, `SessionMetric`, `MetricOptions` and the extension `CreateDriverMetricsProvider` should be on the root namepace. Maybe I should move the other extension methods to the same namespace as `IDriverGauge`, etc.

jorgebay replied:
I think it could be cleaner to move everything to `Metrics` namespace.
There are currently a lot of public members in the root namespace for historical reasons, but we can use correct ones for new features.

joao-r-reis replied:
:+1: moved `SessionMetric, `IMetric`, `NodeMetric` and `MetricOptions` to `Cassandra.Metrics`

jorgebay commented on src/Cassandra/ISession.cs:
```diff
@@ -62,6 +63,12 @@ public interface ISession: IDisposable
+        /// Session name. This can be set with <see cref="Builder.WithSessionName"/>.
+        /// This is used as part of the metric bucket name, for example, which can be used to separate metric paths per session.
+        /// </summary>
+        string SessionName { get; }
```
We could add that it will be autogenerated if not provided.

jorgebay commented on src/Cassandra/Metrics/Registries/INodeMetrics.cs:
```diff
@@ -0,0 +1,59 @@
+
+        IDriverMeter BytesReceived { get; }
+
+        IDriverTimer CqlMessages { get; }
```
`CqlMessages` name for a metric seems off...

joao-r-reis replied:
I agree, I kept the same name as the java driver but I'm ok with renaming it, it's really confusing, do you have a name in mind?

jorgebay replied:
its true that it's the same on the java driver side... nvm

joao-r-reis commented:
I've squashed some commits that were there just to find out what was the bottleneck during benchmarks, you can look at the last commits to see the differences. The major differences are:
- Added an options class for the app metrics provider
- Added HdrReservoir, imported mostly from HdrHistogram.NET and https://github.com/alhardy/AppMetrics.Reservoirs/tree/master/src/App.Metrics.Extensions.Reservoirs.HdrHistogram (unfortunately the app metrics repo doesn't have an up to date package and hdrhistogram.net nuget is not signed so I had to import the relevant code, the licenses offered are CC0 and BSD-2 which I think are ok with this)
- Optimized `Stopwatch.GetTimestamp` calls so that multiple CQL messages on the same "batch" use the same timestamp - `Stopwatch.GetTimestamp` turned out to be a bottleneck after implementing the `HdrReservoir` so this helps a bit.
- Simplified provider interface to make it easier for users to implement a custom provider - now the driver supplies the timer measurement and I've deleted `DriverMeasurementUnit` + `DriverTimeUnit`

joao-r-reis commented:
I'll keep this as `Draft` until I've put together the driver manual section for metrics.

jorgebay commented on src/Cassandra/Requests/RequestExecution.cs:
```diff
@@ -332,28 +337,27 @@ private void HandleException(Exception ex)
-                ex, _parent.RetryPolicy, _parent.Statement, _session.Cluster.Configuration, _retryCount);
-            switch (decision.DecisionType)
+
+            var retryInformation = GetRetryDecisionWithReason(
```
Given that we have the `Unsent` property (connection error before send), we could blindly retry on the next host on those cases.

No need to address it now, we could tackle it in a separate ticket.

joao-r-reis replied:
Hmm we should probably need some more tests on that property before changing behavior based on it. At the moment it's only used to choose which metric counter to increment which is not that critical if the wrong counter is incremented. I'll create a ticket for that.

jorgebay commented on src/Cassandra/Requests/RequestExecution.cs:
```diff
@@ -366,56 +370,94 @@ private void HandleException(Exception ex)
+                    RequestExecution.Logger.Verbose("Socket error " + exception.SocketErrorCode);
+                }
+
+                // For PREPARE requests, retry on next host
```
Just a note, in the future, we could avoid using `RequestHandler` / `RequestExecution` classes for `PREPARE` requests as the behaviour is so different from the rest. We could introduce something like a [`PrepareHandler`](https://github.com/datastax/nodejs-driver/blob/master/lib/prepare-handler.js) :)

jorgebay commented on src/Cassandra/Requests/TcsMetricsRequestResultHandler.cs:
```diff
@@ -0,0 +1,53 @@
+
+namespace Cassandra.Requests
+{
+    internal class TcsMetricsRequestResultHandler : IRequestResultHandler
```
Good idea to introduce this abstraction to wrap this functionality.

jorgebay commented on src/Extensions/Cassandra.AppMetrics/Cassandra.AppMetrics.csproj:
```diff
@@ -0,0 +1,37 @@
+    <Authors>DataStax</Authors>
+    <TargetFrameworks Condition="'$(BuildCoreOnly)' != 'True'">netstandard2.0;net461</TargetFrameworks>
+    <TargetFrameworks Condition="'$(BuildCoreOnly)' == 'True'">netstandard2.0</TargetFrameworks>
+    <NoWarn>$(NoWarn);1591</NoWarn>
```
We should start the project without this `NoWarn` :)

joao-r-reis replied:
:+1:

jorgebay commented on src/Extensions/Cassandra.AppMetrics/HdrHistogram/Histogram.cs:
```diff
@@ -0,0 +1,227 @@
+﻿namespace Cassandra.AppMetrics.HdrHistogram
```
Missing copyright notice.

joao-r-reis replied:
Done

jorgebay commented on src/Extensions/Cassandra.AppMetrics/Implementations/HdrHistogramReservoir.cs:
```diff
@@ -0,0 +1,182 @@
+{
+    /// <summary>
+    /// Adapted code from
+    /// https://github.com/alhardy/AppMetrics.Reservoirs/tree/3a1861386ac73b21b37e548c61575602eed67f89/src/App.Metrics.Extensions.Reservoirs.HdrHistogram
```
I think we must include the original license header.

We can also include a NOTICE file (similar to the one on the core driver) with the same information and linking to the license.

joao-r-reis replied:
Since it's the same repo I added two more "entries" to the already existing NOTICE file.

joao-r-reis commented on src/Cassandra.IntegrationTests/Cassandra.IntegrationTests.csproj:
```diff
@@ -27,13 +28,25 @@
+    <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
+  </ItemGroup>
+  <ItemGroup Condition=" '$(TargetFramework)' == 'net452' ">
+    <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
```
Hmm this was probably changed automatically when I installed `AppMetrics` I guess, nice catch I'll consolidate these.

joao-r-reis commented:
Added manual page for metrics. Added a page for the app metrics provider but it's empty, I'll fill it in tomorrow.

joao-r-reis commented:
Added manual pages for metrics.
Also removed `errors.request.total` because it's tricky to get it right and doesn't have much value (also doesn't exist on the java driver).
Also renamed `errors.requests.*` metrics so they have the same name as the java ones.

This PR is ready to be merged as soon as the review is done. Removing its `Draft` state.

jorgebay commented on doc/features/metrics/README.md:
```diff
@@ -0,0 +1,119 @@
+
+The driver exposes measurements of its internal behavior through an API composed by abstractions. The client application must provide an implementation of these abstractions (`IDriverMetricsProvider`) in order to enable metrics.
+
+DataStax offers an [Extension package] based on the [App.Metrics] library. Application developers can use [App.Metrics] to export metrics to a monitoring tool.
```
TIL that you can use the content of the link as the identifier.

joao-r-reis replied:
Oh I thought I learned that from reading your nodejs docs, maybe it was from the java driver manual I don't remember heh 😄

jorgebay commented on doc/features/metrics/README.md:
```diff
@@ -0,0 +1,119 @@
+
+## Configuration
+
+By default, if `Builder.WithMetrics` is not called by the client application, metrics are disabled.
```
```suggestion
Metrics are disabled by default. To enable them, use `Builder.WithMetrics()` method when creating the session.
```

jorgebay commented on doc/features/metrics/README.md:
```diff
@@ -0,0 +1,119 @@
+
+By default, if `Builder.WithMetrics` is not called by the client application, metrics are disabled.
+
+If `Builder.WithMetrics(provider)` is called, the default metrics will be enabled. The default metrics are every metrics except those of type `Timer`. The reasoning behind this is the fact that enabling `Timer` metrics have the potential effect of reducing the throughput of the driver so it's recommended to benchmark the client application with `Timer` metrics enabled before enabling them in prod ...
```
We could use "When" instead of "If":

```suggestion
When `Builder.WithMetrics(provider)` is called, the default metrics will be enabled. The default metrics are composed of every metric except those of type `Timer`. The reasoning behind this is the fact that enabling `Timer` metrics might increase CPU usage and impact throughput of the driver, so it's recommended to benchmark the client application with `Timer` metrics enabled before enabling them in production.
```

jorgebay commented on doc/features/metrics/README.md:
```diff
@@ -0,0 +1,119 @@
+
+## Retrieving metrics
+
+The driver's API for retrieving metrics is mostly an abstraction. The available methods on each metric type will depend on the provider that is plugged in to the driver.
```
We could recommend to use built-in reporters in the metrics provider to output metrics information (like Graphite/Grafana), noting that "additionally" we expose an API to retrieve it.

joao-r-reis replied:
done

jorgebay commented on doc/features/metrics/app-metrics/README.md:
```diff
@@ -0,0 +1,140 @@
+
+## Installation
+
+This provider is available on the [CassandraCSharpDriver.AppMetrics nuget package].
```
We could publish a `1.0.0-alpha1` ahead of time to reserve the name, otherwise doc generation might fail.

joao-r-reis replied:
Hmm good idea

jorgebay commented on doc/features/metrics/app-metrics/README.md:
```diff
@@ -0,0 +1,140 @@
+
+### Retrieving metrics with the provider's interfaces
+
+The driver's interfaces for each metric type are empty so, in order to be able to retrieve the metric values, we have to retrieve the metric objects with the appropriate provider type.
```
Same as above, we can recommend again to output using a reporter first, I would order the topic of "Exporting" before "Retrieving" :)

joao-r-reis replied:
done

jorgebay commented on doc/features/metrics/metrics-list/README.md:
```diff
@@ -0,0 +1,52 @@
+# List of metrics provided by the driver
```
Awesome!

jorgebay commented on src/Cassandra.IntegrationTests/Cassandra.IntegrationTests.csproj:
```diff
@@ -27,13 +28,25 @@
   </ItemGroup>
+  <ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.0' ">
+    <PackageReference Include="App.Metrics" Version="3.1.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
```
We should move this under the general `<ItemGroup>`.

joao-r-reis replied:
done

joao-r-reis commented:
Pushed changes related to your suggestions.
Also updated api docs project to build api docs for both packages (also changed framework to `netstandard2.0`) and to include descriptions for the new namespaces.

jorgebay commented:
:shipit:
